### DISCUSSION TITLE
How to create a repeatable timer with timeout 0

### DISCUSSION THREAD

**Discussion by Changqing-JING** [libuv/libuv#3432] [Q&A] [answered]
I want to implement a timer like the setInterval(callback, 0) in javascript. But when the uv_timer_start is run with timeout 0 in a timer callback, it will be trigged immediately instead of in next tick.
```c
#include "uv.h"

#include <stdio.h>

uv_timer_t timer1;
uv_loop_t* loop;

void timer_cb1(uv_timer_t* timer) {
	printf("timer_cb1\n");
	uv_timer_start(&timer1, (uv_timer_cb)&timer_cb1, 0, 0);
}

int main() {
	loop = uv_default_loop();
	uv_timer_init(loop, &timer1);
	uv_timer_start(&timer1, (uv_timer_cb)&timer_cb1, 0, 0);

	int ret;

	do{
		printf("tick\n");
		ret = uv_run(loop, UV_RUN_ONCE);
	}while(ret != 0);

	return 0;
}
```

Currently observed behavior:
Keeping print "timer_cb1"

The behavior I want: toggling between "tick" and "timer_cb1"
"tick" -> "tImer_cb1" -> "Tick" -> "timer_cb1".

Can someone kindly provide some idea about how to solve this problem?

**Comment by cjihrig** [comment, marked as answer]
> I want to implement a timer like the setInterval(callback, 0) in javascript.

At least in Node.js, timeouts of 0 are converted to timeouts of 1. According to the `uv_timer_start()` docs:

> If timeout is zero, the callback fires on the next event loop iteration. If repeat is non-zero, the callback fires first after timeout milliseconds and then repeatedly after repeat milliseconds.

I think you want `uv_timer_start(&timer1, (uv_timer_cb)&timer_cb1, 0, 1);`. I don't think you need to call it from `timer_cb1()` either.
